### ISSUE TITLE
Complex app structure

### ISSUE THREAD

**Issue by FabrizioCaldarelli** [yiisoft/yii-core#119]
Starting from basic example 

https://github.com/machour/yii3-kitchen-sink

I have tryed to replicate complex app structure, based on "frontend", "backend" and "api".

So, starting from that example, I'have moved all sources "src" in "src/frontend", thinking that references to sources file were redundant.

![schermata 2018-12-23 alle 12 18 16](https://user-images.githubusercontent.com/4108673/50383066-f8cf7d80-06ac-11e9-8b62-77fbe3d2a0a8.png)

But when I go to http://localhost:8080 I get an exception:

> [Sun Dec 23 12:07:41 2018] An Error occurred while handling another error:
> yii\exceptions\InvalidRouteException: Unable to resolve the request "site/error". in /Users/Fabrizio/Desktop/test2/yii3-kitchen-sink/vendor/yiisoft/yii-core/src/base/Module.php:551
> Stack trace

So, how do you suggest to structure a complex app, based almost on "backend", "frontend" and "api" ?

**Comment by FabrizioCaldarelli** [comment]
Sure, but I only moved all initial "src" folder content (assets, commands, components, controllers, helpers, views) inside a new "src/frontend" folder. 

No other change.

**Comment by machour** [comment]
Well, if you simply move things into another folder, without telling composer/yii where, then it won't work.

You still need to edit composer.json:
(so that autoload knows that idk\app root is in frontend.

```diff
    "autoload": {
-        "psr-4": {"idk\\app\\": "src"}
+        "psr-4": {"idk\\app\\": "src/frontend"}
    },
```

and `config/common.php`

```diff
 return [
     'app' => [
-        'basePath' => dirname(__DIR__) . '/src',
+        'basePath' => dirname(__DIR__) . '/src/frontend',
         'aliases' => [
```

and since you changed composer.json & your config, run `composer du` so that it's taken into account

**Comment by FabrizioCaldarelli** [comment]
This is a bad thing, because framework would depend on composer configuration.

Are you not agree?

**Comment by schmunk42** [comment]
@FabrizioCaldarelli I don't think the framework depends on composer configuration, but your application does.
And do you plan to have two different web entrypoints (`index.php`) for frontend and backend?

**Comment by schmunk42** [comment]
Yes, I think so. I was just thinking about, if developers expect to be able to install a "web-base-backend" or something similar. Would also be interesting how the config plugin can handle this.

But to be honest, I was never a friend of this multiple application tier setup, we always went for the module approach, although I know that developers see this differently in the Yii community.

**Comment by hiqsol** [comment]
> But to be honest, I was never a friend of this multiple application tier setup, we always went for the module approach

That's my point too. Another entry point, another config - it is another application. It has something in common? Put the common part in the new package (module) and use the package in both applications. Of course it leads to creating more packages. But after you get used to it you get only advantages out of it. First of all it forces you write more reusable code.

But! If you still prefer this multiple application approach I think it is achievable.

As for configs you can split them like this:

- config/common/params.php
- config/common/web.php
- config/frontend/web.php
- config/backend/web.php

Then you can merge them with config plugin like this:

```json
    "extra": {
        "config-plugin": {
            "params": "config/common/params.php",
            "frontend": ["config/common/web.php", "config/frontend/web.php"],
            "backend": ["config/common/web.php", "config/backend/web.php"],
        }
    }
```

**Comment by schmunk42** [comment]
> Put the common part in the new package (module) and use the package in both applications. Of course it leads to creating more packages. But after you get used to it you get only advantages out of it. First of all it forces you write more reusable code.

I totally second that! And with composer you can use "private" packages just by including them in the `repositories` section.

**Comment by FabrizioCaldarelli** [comment]
All big applications have almost 2 namespaces : common and backend/frontend/other namespace.

I indeed don't see real benefits in using composer configuration to manage app namespaces, if not create a strong bond between yii and composer.

What about using Yii from archive and not composer? Do we need always to (download and) use composer to make the app working?

I know that it's a bad practice, but when it is possibile I'll reduce every bond with other tools, otherwise Yii will always depends from composer, to work.

In the future tools could be change, and Yii could not choose a new tool because it is linked to composer, for a question that could be solved inside the framework (as it is now).

**Comment by FabrizioCaldarelli** [comment]
I agree with you and I could not imagine a life without package managers.

But

> IMHO at this moment composer is just as basic dependency for php projects as php itself.

in this case this is not true, because we will use composer also to manage namespace structure of the app, as written before.

**Comment by hiqsol** [comment]
Sorry, I didn't get what's the problem with namespaces.

**Comment by FabrizioCaldarelli** [comment]
At the beginning of this issue, my question was about how handle complex app structure.

If I have a complex app with backend, frontend and api, I expect to have sources organized in "src/frontend", "src/backend" and "src/api", within I have models, controllers, views, etc.. such as we do in yii2.

How can I manage this folders structure?

**Comment by machour** [comment]
I'm not sure there's a "recommended" way yet. As I told you, you could have src/frontend, src/backed, or put them both at the root of your project. It's totally up to you as long as you hint composer about what's where. 

Did you try setting something up? I would gladly help you if your face difficulties. I think it's time to try instead of wondering ;)

**Comment by hiqsol** [comment]
It could be something like this:

    api/public/index.php
    api/src/controllers/UserController.php
    backend/public/assets/
    backend/public/index.php
    backend/src/controllers/SettingsController.php
    common/src/Service.php
    frontend/public/assets/
    frontend/public/index.php
    frontend/src/controllers/SiteController.php
    composer.json

Composer.json will hold autoload definition like this:

```json
    "autoload": {
        "psr-4": {
            "complexapp\\common\\": "common/src",
            "complexapp\\api\\": "api/src",
            "complexapp\\backend\\": "backend/src",
            "complexapp\\frontend\\": "frontend/src",
        }
    },
```

`common/src/Service.php` will have `namespace complexapp\common;`

**Comment by terabytesoftw** [comment]
I have the basic application ready, I will focus now on the advanced, I hope the pr of @machour to work perfect with the active record, extension @hiqsol its very simple.

**Comment by FabrizioCaldarelli** [comment]
> Did you try setting something up? I would gladly help you if your face difficulties. I think it's time to try instead of wondering ;)

Yes, I've tried and got this error: 
[Sun Dec 23 12:07:41 2018] An Error occurred while handling another error:
yii\exceptions\InvalidRouteException: Unable to resolve the request "site/error". in /Users/Fabrizio/Desktop/test2/yii3-kitchen-sink/vendor/yiisoft/yii-core/src/base/Module.php:551
Stack trace

@hiqsol So, your example shows that namespaces configuration of complex app (not so much complex, this is a standard app for me) has to be done inside composer.

**Comment by hiqsol** [comment]
> @hiqsol So, your example shows that namespaces configuration of complex app (not so much complex, this is a standard app for me) has to be done inside composer.

Yes. Where else? Yii 3 doesn't provide own class autoloader anymore.

**Comment by FabrizioCaldarelli** [comment]
I'd expect to find namespaces configuration inside the app and not provided by external tools.

**Comment by rob006** [comment]
> I'd expect to find namespaces configuration inside the app and not provided by external tools.

You can always implement your own autoloader and configure autoloading rules in the way you want. Yii dropped its own autoloader because Composer's autoloader is more efficient and flexible, and configuring namespaces in `composer.json` is a standard in the PHP world and most of the people expect to find autoloading configuration right there.

**Comment by FabrizioCaldarelli** [comment]
Perfect, thank you!
